### ISSUE TITLE
How to debug a build that only fails on the new container-based infrastructure?

### ISSUE THREAD

**Issue by halfer** [travis-ci/travis-ci#4785]
I'm using the [Spiderling](https://github.com/OpenBuildings/phpunit-spiderling) test framework with PHPUnit and PhantomJS, and this has worked well on the old Travis infrastructure. I spotted a log message that refers to upgrading to containers, and I don't need `sudo` as far as I know, so I thought I'd try it out.

I added `sudo: false` to my configuration and all my tests involving PhantomJS fail. The web server I spin up (the PHP build-in one) is fine, and Spiderling creates a JavaScript-based server inside PhantomJS to communicate with it, and this clearly works too, as a large log file is still produced.

So, there is something in the container-based infrastructure that is stopping this from working. I've managed to get a screenshot from PhantomJS/Spiderling, but it is just a blank PNG file, suggesting that Phantom has started but is not rendering, or creating a DOM.

I am fresh out of ideas, so I am seeking some community/Travis assistance. I've raised a [Stack Overflow question as well](https://stackoverflow.com/questions/32537787/why-might-phantomjs-fail-to-work-on-the-travis-container-based-infrastructure), and will update here or there, according to where I get answers.

It is possible the bug is in Spiderling, but I can't think what sort of thing it could be doing that would require root permissions (I've never run the tests on Travis, or locally, as root anyway). What other things are different between container and legacy that might cause this?

**Comment by halfer** [comment]
Here is [my working branch](https://github.com/halfer/awooga-app/tree/master) (on legacy) and 
here is the [failing branch](Here is [my working branch](https://github.com/halfer/awooga-app/tree/travis-containers) (new infrastructure). My succeeding and failing builds [can be seen here](https://travis-ci.org/halfer/awooga-app/builds).

**Comment by halfer** [comment]
A fix has come in via Stack Overflow. I'd been setting up the PHP build-in web server to listen on `localhost:port`, which requires a DNS lookup. This is connected to in three places:

* PhantomJS (the `phantomjs` driver)
* cURL inside Spiderling (the `simple` driver)
* A simple server-up test in PHP using `file_get_contents()`

The PhantomJS one seemed to consistently fail, the cURL one seem to always succeed, and the PHP one seemed to succeed but be very slow. Perhaps the internal timeouts in PhantomJS are more strict than elsewhere?

I found that changing the first two to `127.0.0.1` but not the last increased the build time from two minutes to six, and resulted in two failures rather than 20, which would support this theory.

So, this issue is resolved on my side, since I have now changed all listeners and clients to use the dotted numeric address. However, it does not explain why `localhost` is problematic here. I will therefore leave this open for someone to consider, but if it cannot be reproduced, it may be closed.
